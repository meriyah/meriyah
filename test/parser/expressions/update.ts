import { describe } from 'vitest';
import { fail, pass } from '../../test-utils';

describe('Expressions - Update', () => {
  fail('Expressions - Update (fail)', [
    'foo\n++',
    'if (foo\n++);',
    /*['++[]', Context.None],
    ['++([])', Context.None],
    ['(++[])', Context.None],
    ['++[a]', Context.None],
    ['[a]++', Context.None],
    ['[]++', Context.None]*/
  ]);
  pass('Expressions - Update (pass)', [
    'foo\n++\nbar',
    '++\nfoo;',
    'foo\n++bar',
    '++\nfoo;',
    '"foo"\n++bar',
    '+a++ / 1',
    'a=b\n++c',
    'a,b\n++c',
    'a++\nb',
    'a\n++\nb',
    'a.a--',
    '++a.a',
    'foo\n++bar',
    '--a.a',
    '++foo',
    'bar++',
  ]);
});
