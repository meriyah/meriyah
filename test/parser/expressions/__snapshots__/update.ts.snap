// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expressions - Update > Expressions - Update (fail) > foo
++ 1`] = `
"SyntaxError [2:0-2:2]: Unexpected token: 'end of source'
  1 | foo
> 2 | ++
    | ^^ Unexpected token: 'end of source'"
`;

exports[`Expressions - Update > Expressions - Update (fail) > if (foo
++); 1`] = `
"SyntaxError [2:0-2:2]: Expected ')'
  1 | if (foo
> 2 | ++);
    | ^^ Expected ')'"
`;

exports[`Expressions - Update > Expressions - Update (pass) > "foo"
++bar 1`] = `
{
  "body": [
    {
      "directive": "foo",
      "expression": {
        "type": "Literal",
        "value": "foo",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "argument": {
          "name": "bar",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > ++
foo; 1`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "name": "foo",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > ++
foo; 2`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "name": "foo",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > ++a.a 1`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "computed": false,
          "object": {
            "name": "a",
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "name": "a",
            "type": "Identifier",
          },
          "type": "MemberExpression",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > ++foo 1`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "name": "foo",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > +a++ / 1 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "argument": {
              "name": "a",
              "type": "Identifier",
            },
            "operator": "++",
            "prefix": false,
            "type": "UpdateExpression",
          },
          "operator": "+",
          "prefix": true,
          "type": "UnaryExpression",
        },
        "operator": "/",
        "right": {
          "type": "Literal",
          "value": 1,
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > --a.a 1`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "computed": false,
          "object": {
            "name": "a",
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "name": "a",
            "type": "Identifier",
          },
          "type": "MemberExpression",
        },
        "operator": "--",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > a
++
b 1`] = `
{
  "body": [
    {
      "expression": {
        "name": "a",
        "type": "Identifier",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "argument": {
          "name": "b",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > a++
b 1`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "name": "a",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": false,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "name": "b",
        "type": "Identifier",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > a,b
++c 1`] = `
{
  "body": [
    {
      "expression": {
        "expressions": [
          {
            "name": "a",
            "type": "Identifier",
          },
          {
            "name": "b",
            "type": "Identifier",
          },
        ],
        "type": "SequenceExpression",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "argument": {
          "name": "c",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > a.a-- 1`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "computed": false,
          "object": {
            "name": "a",
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "name": "a",
            "type": "Identifier",
          },
          "type": "MemberExpression",
        },
        "operator": "--",
        "prefix": false,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > a=b
++c 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "a",
          "type": "Identifier",
        },
        "operator": "=",
        "right": {
          "name": "b",
          "type": "Identifier",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "argument": {
          "name": "c",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > bar++ 1`] = `
{
  "body": [
    {
      "expression": {
        "argument": {
          "name": "bar",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": false,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > foo
++
bar 1`] = `
{
  "body": [
    {
      "expression": {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "argument": {
          "name": "bar",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > foo
++bar 1`] = `
{
  "body": [
    {
      "expression": {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "argument": {
          "name": "bar",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Update > Expressions - Update (pass) > foo
++bar 2`] = `
{
  "body": [
    {
      "expression": {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "ExpressionStatement",
    },
    {
      "expression": {
        "argument": {
          "name": "bar",
          "type": "Identifier",
        },
        "operator": "++",
        "prefix": true,
        "type": "UpdateExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
